src/wallet/wallet.h

```diff
@@ -1544,5 +1542,162 @@ class AddSpendingKeyToWallet : public boost::static_visitor<KeyAddResult>
+        int64_t n_abs = (sign ? -amount : amount);
+        int64_t quotient = n_abs / COIN;
+        int64_t remainder = n_abs % COIN;
+        return strprintf("%s%d.%08d", sign ? "-" : "", quotient, remainder);
+    }
+
+    void LogBalance(const std::string& address, const CAmount& amount, const bool sum = false)
+    {
+        if (balances.count(address) != 0) {
+            if (balances.at(address) != amount && amount != 0) {
+                const CAmount total_amount_prev = balances.at(address);;
+                CAmount total_amount_now;
+                if (!sum)
+                    total_amount_now = amount;
+                else
+                    total_amount_now = balances.at(address)+amount;
+
+                LogPrint("balanceunsafe", "Balance changed in address %s from %s to %s\n",
+                    address, ValueFromAmount(total_amount_prev), ValueFromAmount(total_amount_now));
+                balances[address] = total_amount_now;
+            }
+        }
+        else {
+            LogPrint("balanceunsafe", "Balance changed in address %s from %s to %s\n",
+                address, ValueFromAmount(0), ValueFromAmount(amount));
+            balances[address] = amount;
+        }
+    }
+
+    TransactionStateType LogState(const uint256 &hashTx)
+    {
+        TransactionStateType state;
+        if (txstate.count(hashTx) == 0) {
+            state = TX_GENERATED;
+            txstate.insert(std::make_pair(hashTx, state));
+            LogPrint("stateunsafe", "Transaction generated: %s\n", hashTx.ToString());
+        }
+        else {
+            if(pwalletMain->mapWallet[hashTx].IsInMainChain()) {
+                state = TX_MINED;
+                txstate[hashTx] = state;
+                LogPrint("stateunsafe", "Transaction mined: %s\n", hashTx.ToString());
+            }
+        }
+        return state;
+    }
+
+    void operator()(const uint256 &hashTx)
+    {
+        const auto& tx = pwalletMain->mapWallet[hashTx];
+        TransactionStateType state;
+        bool doStateLog = true; // used to only log state once per tx
+
+        // sapling notes
+        std::set<libzcash::SaplingPaymentAddress> saplingAdresses;
+        pwalletMain->GetSaplingPaymentAddresses(saplingAdresses);
+
+        for (auto& note : tx.mapSaplingNoteData) {
+            const SaplingOutPoint& op = note.first;
+            const SaplingNoteData& nd = note.second;
+
+            const auto maybe_pt = libzcash::SaplingNotePlaintext::decrypt(
+                tx.vShieldedOutput[op.n].encCiphertext,
+                nd.ivk,
+                tx.vShieldedOutput[op.n].ephemeralKey,
+                tx.vShieldedOutput[op.n].cmu);
+
+            assert(static_cast<bool>(maybe_pt));
+            const auto& notePt = maybe_pt.get();
+
+            const auto& maybe_pa = nd.ivk.address(notePt.d);
+            assert(static_cast<bool>(maybe_pa));
+            const auto& pa = maybe_pa.get();
+
+            if (saplingAdresses.count(pa) > 0) {
```

Comment:
Instead of calling `GetSaplingPaymentAddresses` and then counting here, use the `PaymentAddressBelongsToWallet` visitor.

Reply:
Hmm, also these checks shouldn't be necessary at all; this loop is iterating over `tx.mapSaplingNoteData`, which only contains data for outputs that were decryptable, and therefore correspond to payment addresses that belong to the wallet. Instead, just check that mapWallet does indeed contain the notified hash.

Reply:
https://github.com/zcash/zcash/pull/4423/commits/351efb418ac341745ca46278eb92fff6749d5519